### ISSUE TITLE
Custom batches for training with in-batch negatives (MultipleNegativesRankingLoss)

### ISSUE THREAD

**Issue by lmores** [UKPLab/sentence-transformers#3743]
I am using this library to finetune a model. The goal is to generate embeddings for postal addresses so that embeddings for the same address written in different manners are close to each other.
However, addresses that only differ for a small part (e.g. the street number, or the name of the city) must have sufficiently different embeddings, which is not the case when I try to finetune the `all-mpnet-base-v2` model using the `CosineSimilarityLoss` (providing both positive and negative pairs).

Therefore, I am trying to use the [`MultipleNegativesRankingLoss`](https://sbert.net/docs/package_reference/sentence_transformer/losses.html#multiplenegativesrankingloss). As far as I understand, the computation of this loss function takes into account the whole minibatch, not just the individual pairs of sentences/addresses. It enforces not only that sentences/addresses in a given pair have similar embeddings, but also consider sentences/addresses in different pairs of the same batch as negatives (which is exactly what I need).

Therefore, I prepared a trainining set that is already partitioned in batches with 256 pairs each, taking care to put in the same batch pairs that must be considered strong negatives even if they are quite similar.

```python
batches: list[tuple[tuple[str, str], 256]] = [
    (
        (batch1_anchor1,  batch1_positive1),  # ('Blue Street, 1, New York', 'Blue Street 1 - New York'), 
        (batch1_anchor2,  batch1_positive2),  # ('Blue Street, 11, New York', 'Blue Street 11 - New York'),
        (batch1_anchor3,  batch1_positive3),
        ...
    ),
    (
        (batch2_anchor1,  batch2_positive1),
        (batch2_anchor2,  batch2_positive2),
        (batch2_anchor3,  batch2_positive3),
        ...
    ),
    ....
]
```

My question is: how do I preserve this batch structure when loading the training data into the trainer?
The [`SentenceTransformerTrainer`](https://sbert.net/docs/package_reference/sentence_transformer/trainer.html#sentencetransformertrainer) class only accepts a `datasets.Dataset`, I see no way to preserve my batches.

```python
    loss_fn = MultipleNegativesRankingLoss(
        model, 
        directions=('query_to_doc', 'query_to_query', 'doc_to_query', 'doc_to_doc')
    )

    trainer = SentenceTransformerTrainer(
        model=model,
        args=args,
        train_dataset=???,  # here I can pass a datasets.Dataset, not a torch.utils.data.DataLoader or equivalent
        loss=loss_fn,
     )
```

Note: also asked on [HugginFace forum](https://discuss.huggingface.co/t/custom-batches-in-sentence-transformers-for-multiplenegativesrankingloss/175587).

**Comment by tomaarsen** [comment]
Hello!

Thanks for the detailed write-up, and sorry for the delayed response. Your understanding of `MultipleNegativesRankingLoss` is correct: it treats every other positive in the batch as an in-batch negative, so the *composition* of the batch directly shapes what the model learns to separate. Putting `"Blue Street 1, New York"` and `"Blue Street 11, New York"` in the same batch is exactly the kind of hard signal MNRL benefits from.

Two paths forward, plus one note on your loss config:

### Custom batch sampler

`SentenceTransformerTrainingArguments` accepts a `batch_sampler` argument that takes either a `DefaultBatchSampler` subclass or a factory function. Flatten your pre-grouped batches into a `Dataset` while preserving row order, then write a small sampler that yields the pre-defined index groups:

```python
import torch
from datasets import Dataset
from sentence_transformers import (
    SentenceTransformer,
    SentenceTransformerTrainer,
    SentenceTransformerTrainingArguments,
)
from sentence_transformers.base.sampler import DefaultBatchSampler
from sentence_transformers.losses import MultipleNegativesRankingLoss

# Flatten your pre-grouped batches into ordered rows.
anchors, positives = [], []
for batch in batches:
    for a, p in batch:
        anchors.append(a)
        positives.append(p)
train_dataset = Dataset.from_dict({"anchor": anchors, "positive": positives})


class PredefinedBatchSampler(DefaultBatchSampler):
    """Yields contiguous index blocks of size ``batch_size``, preserving the
    pre-curated batch composition. Batch *order* is reshuffled per epoch.
    """

    def __iter__(self):
        if self.generator and self.seed is not None:
            self.generator.manual_seed(self.seed + self.epoch)
        num_batches = len(self.dataset) // self.batch_size
        order = torch.randperm(num_batches, generator=self.generator).tolist()
        for b in order:
            samples = list(range(b * self.batch_size, (b + 1) * self.batch_size))
            print(samples)  # Keep only for debugging to verify batch composition
            breakpoint()
            yield samples

    def __len__(self):
        return len(self.dataset) // self.batch_size


model = SentenceTransformer("sentence-transformers/all-mpnet-base-v2")
loss_fn = MultipleNegativesRankingLoss(model)

args = SentenceTransformerTrainingArguments(
    output_dir="checkpoints",
    per_device_train_batch_size=256,  # must match your pre-curated batch size exactly
    batch_sampler=PredefinedBatchSampler,
    dataloader_drop_last=True,
)

trainer = SentenceTransformerTrainer(
    model=model,
    args=args,
    train_dataset=train_dataset,
    loss=loss_fn,
)
trainer.train()
```

A couple of things to watch:

* `per_device_train_batch_size` must equal the size of your pre-curated batches (256 in your example).
* On multi-GPU, each rank's in-batch negatives are local to that rank by default, so the effective negatives per anchor are `per_device_train_batch_size - 1`, not `world_size * per_device_train_batch_size - 1`. If you want global negatives across ranks, `CachedMultipleNegativesRankingLoss` is the cleaner path.

I added a small print/breakpoint that you can use to double-check that the sampler is yielding the expected batches.

### Explicit hard-negative columns

Instead of curating batches, you can also keep one row per query with positives and hard negatives as separate columns: `(anchor, positive, negative_1, negative_2, ...)`. `MultipleNegativesRankingLoss` natively uses each row's own negatives in addition to in-batch negatives, so no sampling tricks are needed.

`sentence_transformers.util.mine_hard_negatives` finds near-duplicate addresses for each query using a base embedder. This is essentially what you're doing manually, but global rather than per-batch:

```python
from sentence_transformers.util import mine_hard_negatives

mined = mine_hard_negatives(
    dataset=train_dataset,    # (anchor, positive) pairs
    model=model,
    range_min=10,             # skip top-10 most similar (likely true matches)
    range_max=50,             # take from positions 10..50
    max_score=0.95,           # avoid near-identical paraphrases
    relative_margin=0.05,     # negative is at most 95% as similar as the positive
    num_negatives=5,
    output_format="n-tuple",  # one row per anchor with all negatives as columns
)
# mined now has columns: anchor, positive, negative_1, ..., negative_5
```

The `range_min` / `range_max` / `relative_margin` / `max_score` knobs let you target the "similar but not identical" band where street-number and city swaps live, which is exactly your problem. This composes with normal shuffling, scales better than hand-curating, and is what most users converge on for this kind of task.

However, you'll still get in-batch negatives that you might've wanted to have as positives. If that's a problem, you'd have to use a custom loss like TripletLoss (but that accepts more than 1 negative per anchor).

Lastly, a note on your loss config:

```python
MultipleNegativesRankingLoss(
    model,
    directions=("query_to_doc", "query_to_query", "doc_to_query", "doc_to_doc"),
)
```

These directions are totally fine, but depending on your actual batch composition, you might want to exclude some. It should be simple enough to test if you add an evaluator like `TripletEvaluator` or `BinaryClassificationEvaluator`.

- Tom Aarsen

**Comment by lmores** [comment]
Thank you for your very detailed answer. I am using your `PredefinedBatchSampler` class and everything seems to work smoothly. I just had to replace `self.dataset` with `self.sampler` as the `DefaultBatchSampler` class passes the provided `dataset` as the first argument to its parent `__init__` method:

https://github.com/huggingface/sentence-transformers/blob/4d2d06c871a17ee83b4984aa4e4b37638e15741c/sentence_transformers/base/sampler.py#L220

and it is stored in `self.sampler` by [`torch.utils.data.BatchSampler.__init__`](https://github.com/pytorch/pytorch/blob/70d99e998b4955e0049d13a98d77ae1b14db1f45/torch/utils/data/sampler.py#L308-L331). I am not sure if this name mismatch is wanted or not.

If you do not mind, I have a couple more questions based on your answer.

A) As you mentioned:

> Instead of curating batches, you can also keep one row per query with positives and hard negatives as separate columns: (anchor, positive, negative_1, negative_2, ...). MultipleNegativesRankingLoss natively uses each row's own negatives in addition to in-batch negatives, so no sampling tricks are needed.

When you say "no sampling tricks are needed", are you assuming that variants of the same address do not appear in different `(anchor, positive)` pairs in the dataset? Otherwise they may be (randomly) put in the same batch and give a false signal, is it correct? (This is actually may case).

Also, if I add `negative_1, ..., negative_N` to my pairs, I must provide the same number N of negatives for each anchor, right? (This is also something I cannot do easily, as I assembled my training dataset using various sources and, depending on the source, I may or may not be able to easily generates hard negatives for the anchors).

I read about the `mine_hard_negatives` function, but I did not believe it could actually make a better job than what I could achieve manually (although taking much more time), maybe I'll give it a shot if I will need to rebuild my training dataset, thank you!

B) Regarding:

> These directions are totally fine, but depending on your actual batch composition, you might want to exclude some. It should be simple enough to test if you add an evaluator like TripletEvaluator or BinaryClassificationEvaluator.

Thanks for the hint, I did use all 4 directions as my `(anchor, positive)` pair are totally symmetric and unique within a batch. Is this choice correct?

Also, I am bit confused about how you suggests to check if some of them are redundant. Did you mean _"just remove one or two directions and check how the new finedtuned model performs using an evaluator"_ or something different?

**Comment by tomaarsen** [comment]
> Thank you for your very detailed answer. I am using your `PredefinedBatchSampler` class and everything seems to work smoothly. I just had to replace `self.dataset` with `self.sampler` as the `DefaultBatchSampler` class passes the provided `dataset` as the first argument to its parent `__init__` method:
> 
> [sentence-transformers/sentence_transformers/base/sampler.py](https://github.com/huggingface/sentence-transformers/blob/4d2d06c871a17ee83b4984aa4e4b37638e15741c/sentence_transformers/base/sampler.py#L220)
> 
> Line 220 in [4d2d06c](/huggingface/sentence-transformers/commit/4d2d06c871a17ee83b4984aa4e4b37638e15741c)
>  super().__init__(dataset, batch_size=batch_size, drop_last=drop_last) 
> 
> and it is stored in `self.sampler` by [`torch.utils.data.BatchSampler.__init__`](https://github.com/pytorch/pytorch/blob/70d99e998b4955e0049d13a98d77ae1b14db1f45/torch/utils/data/sampler.py#L308-L331). I am not sure if this name mismatch is wanted or not.

Ah, yes, `self.sampler` should do it. I pass `dataset` as I feel it's more indicative of what it is in our case, but the `BatchSampler` superclass is broader and uses `self.sampler` instead.

> When you say "no sampling tricks are needed", are you assuming that variants of the same address do not appear in different (anchor, positive) pairs in the dataset? Otherwise they may be (randomly) put in the same batch and give a false signal, is it correct? (This is actually may case).

Yes. Admittedly, it wasn't the best suggestion, because in your dataset there's a lot of risk for accidental false negatives if you go this hard negatives mining route.

> Also, if I add `negative_1, ..., negative_N` to my pairs, I must provide the same number N of negatives for each anchor, right? (This is also something I cannot do easily, as I assembled my training dataset using various sources and, depending on the source, I may or may not be able to easily generates hard negatives for the anchors).

That's right, I'm afraid. For bi-encoder models, the number of negatives must be the same for every query. This is a common annoyance, also for myself, but it's a bit tricky to rework at this time. As for `mine_hard_negatives`: I would see it as a time-saver or convenience thing. If you can manually find the positives and negatives, then that's certainly preferable. 

> Thanks for the hint, I did use all 4 directions as my `(anchor, positive)` pair are totally symmetric and unique within a batch. Is this choice correct?
> 
> Also, I am bit confused about how you suggests to check if some of them are redundant. Did you mean "just remove one or two directions and check how the new finedtuned model performs using an evaluator" or something different?

"Correct" is hard to say, there's a bit of magic to training models that makes it tricky to know in advance which will work best. And yes, I do mean that you can remove some directions (except the required `query_to_doc`) and see how the new finetuned model works on your evaluator. I would try at least these 3:
1. `query_to_doc`, `doc_to_query`, `query_to_query`, `doc_to_doc`
2. `query_to_doc`, `doc_to_query`
3. `query_to_doc`

In your case, I suspect that 1 or 2 will be best, while 3 is most commonly used in the literature. I'm glad to hear that the training is progressing nicely!

- Tom Aarsen
